pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -288,6 +331,313 @@ extension type Schema.fromMap(Map<String, Object?> _value) {
+    for (final subSchema in schema.anyOf!) {
+      if (_validateSchema(subSchema, data).isEmpty) {
+        oneValid = true;
+        break;
+      }
+    }
+    if (!oneValid) {
+      failures.add(ValidationFailureType.anyOfNotMet);
+    }
+  }
+  if (schema.oneOf != null) {
+    var validCount = 0;
+    for (final subSchema in schema.oneOf!) {
+      if (_validateSchema(subSchema, data).isEmpty) {
+        validCount++;
+      }
+    }
+    if (validCount != 1) {
+      failures.add(ValidationFailureType.oneOfNotMet);
+    }
+  }
+  if (schema.not != null) {
+    var validCount = 0;
+    for (final subSchema in schema.not!) {
+      if (_validateSchema(subSchema, data).isEmpty) {
+        validCount++;
+      }
+    }
+    if (validCount != 1) {
+      failures.add(ValidationFailureType.notConditionViolated);
+    }
+  }
+
+  // 2. Handle explicit type validation if schema.type is present
+  final schemaType = schema.type;
+  if (schemaType != null) {
+    switch (schemaType) {
+      case JsonType.object:
+        if (data is Map<String, Object?>) {
+          failures.addAll(_validateObject(schema as ObjectSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.list:
+        if (data is List) {
+          failures.addAll(_validateList(schema as ListSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.string:
+        if (data is String) {
+          failures.addAll(_validateString(schema as StringSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.num:
+        if (data is num) {
+          failures.addAll(_validateNumber(schema as NumberSchema, data));
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.int:
+        if (data is int) {
+          failures.addAll(_validateInteger(schema as IntegerSchema, data));
+        } else if (data is num && data == data.toInt()) {
+          failures.addAll(
+            _validateInteger(schema as IntegerSchema, data.toInt()),
+          );
+        } else {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.bool:
+        if (data is! bool) {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+      case JsonType.nil:
+        if (data != null) {
+          failures.add(ValidationFailureType.typeMismatch);
+        }
+        break;
+    }
+  }
+
+  return failures.toSet().toList(); // Remove duplicates
+}
+
+List<ValidationFailureType> _validateObject(
+  ObjectSchema schema,
+  Map<String, Object?> data,
+) {
+  final failures = <ValidationFailureType>[];
+
+  if (schema.minProperties != null &&
+      data.keys.length < schema.minProperties!) {
```

Comment:
I would use if-case throughout these as well, same for the list validation below

```suggestion
  if (schema.minProperties case final min? when data.keys.length < min) {
```